packages/go_router/example/lib/stateful_nested_navigation.dart

```diff
@@ -0,0 +1,291 @@
+              /// The details screen to display stacked on navigator of the
+              /// second tab. This will cover screen B but not the application
+              /// shell (bottom navigation bar).
+              GoRoute(
+                path: 'details/:param',
+                builder: (BuildContext context, GoRouterState state) =>
+                    DetailsScreen(label: 'B', param: state.params['param']),
+              ),
+            ],
+          ),
+        ],
+
+        /// If you need to customize the Page for StatefulShellRoute, pass a
+        /// pageProvider function in addition to the builder, for example:
+        // pageProvider:
+        //     (BuildContext context, GoRouterState state, Widget statefulShell) {
+        //   return NoTransitionPage<dynamic>(child: statefulShell);
+        // },
+
+        /// If you need to create a custom container for the branch routes, to
+        /// for instance setup animations, you can implement your builder
+        /// something like this:
+        // builder:
+        //     (BuildContext context, GoRouterState state, Widget ignoringThis) {
+        //   final StatefulShellRouteState shellRouteState =
+        //       StatefulShellRoute.of(context);
+        //   return ScaffoldWithNavBar(
+        //       body: _AnimatedRouteBranchContainer(
+        //     currentIndex: shellRouteState.index,
+        //     navigators: shellRouteState.navigators,
+        //   ));
+        // },
+      ),
+    ],
+  );
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp.router(
+      title: 'Flutter Demo',
+      theme: ThemeData(
+        primarySwatch: Colors.blue,
+      ),
+      routerConfig: _router,
+    );
+  }
+}
+
+/// Builds the "shell" for the app by building a Scaffold with a
+/// BottomNavigationBar, where [child] is placed in the body of the Scaffold.
+class ScaffoldWithNavBar extends StatelessWidget {
+  /// Constructs an [ScaffoldWithNavBar].
+  const ScaffoldWithNavBar({
+    required this.body,
+    Key? key,
+  }) : super(key: key ?? const ValueKey<String>('ScaffoldWithNavBar'));
+
+  /// Body, i.e. the index stack
+  final Widget body;
+
+  @override
+  Widget build(BuildContext context) {
+    final StatefulShellRouteState shellState = StatefulShellRoute.of(context);
+    return Scaffold(
+      body: body,
+      bottomNavigationBar: BottomNavigationBar(
+        items: const <BottomNavigationBarItem>[
+          BottomNavigationBarItem(icon: Icon(Icons.home), label: 'Section A'),
+          BottomNavigationBarItem(
+              icon: Icon(Icons.settings), label: 'Section B'),
+        ],
+        currentIndex: shellState.index,
+        onTap: (int tappedIndex) => _onItemTapped(
+          context,
+          shellState.navigationBranchState[tappedIndex],
+        ),
+      ),
+    );
+  }
+
+  void _onItemTapped(BuildContext context, ShellRouteBranchState routeState) {
+    GoRouter.of(context).go(routeState.location);
```

Comment:
Yes, you're right - or rather it wouldn't make sense to push a route from one route branch onto another branch. That is, you should only be allowed to push a route if it's a descendant of the same branch as the current one (if any). And there is of course nothing that prevents anyone from attempting to push a location under a StatefulShellRoute from anywhere, so the validation (trow exception) on push might still be needed. 

The api for going to a specific route index would also be good of course, although I'm trying to decide if routeState.location could still be useful (or needed). But I can't come up with such a use case right now, so I guess it might be better to hide it for now, and expose it later if the need should arise.

Reply:
@chunhtai, do you have a suggestion for an appropriate name of the method in `StatefulShellRouteState` for navigating to another branch index? Right now I'm calling it `goToBranch`, but would `goBranch` be more in line with what the methods are called in `GoRouter` (i.e. `goNamed`)? Another candidate is perhaps `switchBranch`.

Reply:
Done.